tutorial/ndarrays.md

```diff
@@ -0,0 +1,328 @@
+# The N-dimensional array (ndarray) #
+
+## Description ##
+
+A ndarray is a fixed-size multi-dimensional container of items of the same type and size. The number of dimensions and items in an array is defined by its shape, which is a container of N non-negative integers that specify the sizes of each dimension. The type of items in the array is specified by a separate data-type object, one of which is associated with each ndarray.
+
+Different ndarrays can share the same data, so that changes made in one ndarray may be visible in another. That is, a ndarray can be a "view" to another ndarray, and the data it is referring to is taken care of by the "base" ndarray.
+[read more](https://numpy.org/doc/stable/reference/arrays.html)
+
+## Pyccel ndarrays ##
+
+Pyccel uses the same implementation as Numpy ndarrays with some rules due to the difference between the host language (Python) "dynamically typed / internal garbage collector" and the target languages such as C and Fortran which are statically typed languages and don't have a garbage collector.
+
+Below we will show some rules that Pyccel has set to handles those differences.
+
+### Dynamically and statically typed languages ###
+
+Generally a variable in Pyccel should always keep its initial type, this also transfers to using the ndarrays.
+
+#### incorrect example ####
+
+```Python
+import numpy as np
+
+a = np.array([1, 2, 3], dtype=float)
+#(some code...)
+a = np.array([1, 2, 3], dtype=int)
+```
+
+_OUTPUT_ :
+
+```Shell
+ERROR at annotation (semantic) stage
+pyccel:
+ |error [semantic]: ex.py [5]| Incompatible types in assignment (|a| real <-> int)
+```
+
+### Memory management ###
+
+Pyccel makes a difference between ndarrays that own their data and the ones they don't.
+
+Pyccel calls its own garbage collector when needed, but has a set of rules to do so:
+
+-   Can not reassign ndarrays with different ranks.
+
+    ```Python
+    import numpy as np
+
+    a = np.ones((10, 20))
+    #(some code...)
+     a = np.ones(10)
+    ```
+
+    _OUTPUT_ :
+
+    ```Shell
+    ERROR at annotation (semantic) stage
+    pyccel:
+     |error [semantic]: ex.py [4]| Incompatible redefinition (|a| real(10, 20) <-> real(10,))
+    ```
+
+    This limitation is due to the way Fortran allocatable can't change the rank after declaration.
+
+-   Can not assign ndarrays that own their data one another.
+
+    ```Python
+    import numpy as np
+
+    a = np.array([1, 2, 3, 4, 5])
+    b = np.array([1, 2, 3])
+    a = b
+    ```
+
+    _OUTPUT_ :
+
+    ```Shell
+    ERROR at annotation (semantic) stage
+    pyccel:
+     |error [semantic]: ex.py [5]| Arrays which own their data cannot become views on other arrays (a)
+    ```
+
+    This limitation is due to the fact that the ndarray **a** will have to go from a data owner to a pointer to the **b** ndarray data.
+
+    _NOTE_: this limitation does not include reassigning using new data with respecting the previous rule.
```

Comment:
I don't understand what you mean here

Reply:
i tried to say that the limitation of reassigning a ndarray variable is not in place when the right-hand side is a newly created array (reallocation).

Reply:
Ah, you mean:
```
This limitation only applies to data that is shared between arrays, not to assignments which reserve a new memory block
```
?

Reply:
That sounds clearer :+1: